### ISSUE TITLE
sendBetterEmail - An Apex error occurred: SendBetterEmail.InvocableActionException: No Messages were sent.  First Error: nullError Code:INVALID_EMAIL_ADDRESS - Email address is invalid: 

### ISSUE THREAD

**Issue by Mitchelldevos** [alexed1/LightningFlowComponents#663]
When I try to use the SendBetterEmail component within a flow and I don't fill some optional fields like Sender Display Name, Reply Email Address, CC address, BCC address I get the following error:

FLOW_ELEMENT_ERROR An Apex error occurred: SendBetterEmail.InvocableActionException: No Messages were sent.  First Error: nullError Code:INVALID_EMAIL_ADDRESS - Email address is invalid: 

The email I use, within the ToAddress is valid, so that's not it.
It looks like it doesn't accept leaving the Sender Display Name, Reply Email Address, CC address, BCC address fields empty. 

This is since a week or so, before it just worked with only filling in the ToAddress and leaving the Sender Display Name, Reply Email Address, CC address, BCC address fields empty. 

![SendBetterEmail - Invalid Email Address Error](https://user-images.githubusercontent.com/33150125/112468738-ef8c5a00-8d68-11eb-9781-9591c86e1e4e.png)

**Comment by Mitchelldevos** [comment]
Extra input.

<img width="1153" alt="SendBetterEmail - Flow" src="https://user-images.githubusercontent.com/33150125/112653760-09a06800-8e4f-11eb-8268-714ad15eff93.png">
<img width="1920" alt="SendBetterEmail - Config 1" src="https://user-images.githubusercontent.com/33150125/112653771-0c02c200-8e4f-11eb-9f58-d6487f51d6ed.png">
<img width="1920" alt="SendBetterEmail - Config 2" src="https://user-images.githubusercontent.com/33150125/112653780-0dcc8580-8e4f-11eb-99eb-16407aaa5807.png">

**Comment by jdpond** [comment]
when did you upgrade to 2.1.0 and when was the last time you saved this flow?

**Comment by Mitchelldevos** [comment]
I installed version 2.1.0 so didn't upgrade. Didn't have this version before.
I made some changes and saved this flow yesterday. 

It only works for me if I also fill in Sender Display Name, Reply Email Address, CC address, BCC address.

**Comment by jdpond** [comment]
Thanks for helping me debug this.  Could you also remove the CC Address and BCC address?  and tell me what you get?

**Comment by Mitchelldevos** [comment]
When I remove CC and BCC address I got the same error.

**Comment by Rick131313** [comment]
I am experiencing a similar issue...same error message.  Mine just started yesterday because it had been working without me having to populate cc or bcc.  So I reconfigured my email address collections/variables and tried it using a single email address in the To line and a collection of emails in the CC line.  Same error.  I too just installed version 2.1.0 as my first version.  Then I removed the CC entry and put a collection of emails variable in as the To and still getting the exact error as above.

**Comment by Rick131313** [comment]
I saw a screen of Input Values which appears to be on the SendHTMLEmails version.  I got excited when I saw it because I thought it was a screen I must have missed because it is exactly what I was looking for...a way to turn on just those email features you need. Is that ability still out there and I just missed it?

**Comment by Rick131313** [comment]
I decided to create a simple flow from scratch which just sends an HTML body email (one liner) using three different sendBetterEmail envelope options: 1) To: (text variable with one email address entered into the Default box)  2) To: (string collection variable containing one text variable which was populated with one email address using "Add" in an Assignment element)  3) To: (used a collection variable containing two variables [a string variable and a collection variable].

All three of the actions worked perfectly.  Here's what I noticed that was different:  When I pulled down the drop down to set the To: field, it said "Select an option".  I believe this is critical because if "Select an option" is the active option, the Apex action seems to know not to look for an email address in that field (this effects cc or bcc).  Once something has been entered and the "Select an option" option is no longer an option in the pulldown, that's when things start going sideways.

Is "Select an option" supposed to be the option that signals the SBE action not to expect an email address in that field so it skips it?  That is the best theory I can come up with.  I don't know how to read code so I can't tell if that is how it's coded or not.

**Comment by mpaul1** [comment]
I was able to work around this issue by populating the CC with a properly formatted but bogus email address.  But, since our org has BCC compliance enabled, doing the same for the BCC field is not an option.  I was able to fix this by removing the following map entries(Below) in the SendBetterEmailUtil class and leaving BCC blank.  Not ideal, but it works...

'SendBCCthisOneEmailAddress' => request.SendBCCthisOneEmailAddress,
'SendBCCthisStringCollectionOfEmailAddresses' => request.SendBCCthisStringCollectionOfEmailAddresses,
'SendBCCtheEmailAddressesFromThisCollectionOfContacts' => request.SendBCCtheEmailAddressesFromThisCollectionOfContacts,
'SendBCCtheEmailAddressesFromThisCollectionOfUsers' => request.SendBCCtheEmailAddressesFromThisCollectionOfUsers,
'SendBCCtheEmailAddressesFromThisCollectionOfLeads' => request.SendBCCtheEmailAddressesFromThisCollectionOfLeads

**Comment by Rick131313** [comment]
Michael, thanks for the follow-up.  I found that if I made edits to the To: section of the SendBetterEmails Alert (that's basically what the Apex Action is...an Email Alert) it was like the form didn't forget what was there before and just because you overwrote it doesn't mean it forgot it.  So, once I figured that out (by the way, this was the solution for many issues with the SendBetterEmails Alert form) I would delete the Apex Action that seemed to be causing the problem (look at the debug report) and re-create it knowing exactly what you want in the To:, CC:, and BCC: fields.  I populated these three envelope fields with string collection variables which I had stored all the email addresses in.  I never used a direct email address in the SBE envelope...just variables.  That way you don't have to change what's on the form...just change what's in the variable collection via Assignments.

It took a bit of testing to figure out how the SBE Alert form worked but once I did, all was good.

Rick

**Comment by jdpond** [comment]
I am going to leave this one open because it points to a bigger issue - even though @Rick131313 has come up with a generic and very useful workaround to be used until this is remediated (delete old action and insert new from scratch).   We need to re-introduce logic flow - including the ability to error check and/or ignore irrelevant data previously entered.  Any volunteers to help me with this process would be greatly appreciated.

TL;DR - when we refactored the component we eliminated the linear logic flow paths in favor of mapped asynchronous logic which made the code "architecturally attractive" - but at a cost.  Object lesson -"Pluralitas non est ponenda sine neccesitate(entities should not be multiplied unnecessarily)" [Also known as Ockham's Razor] --William of Ockham(~1350)

**Comment by Rick131313** [comment]
Jack, thanks for the follow-up...truly appreciated!  Just to add on to my experience:

As mentioned, I replaced all the actions with new actions and everything was fine.  I was using collection variables to store the TO email addresses and all worked perfectly.  Then, the other day someone asked if we could blind copy the sending email address being used to send the email (same as ReplyToEmailAddress).  So I used the variable containing the ReplyToEmailAddress and placed it inside a collection variable, EmailAddressesToBCC (or something like that).  Good news is the bcc worked just fine.  Unfortunately however, I am now receiving a copy of every email being sent via the action in my work inbox.  Of course when you review the "metadata" of the email, it only lists the TO email addresses (I'm not one of them) and the BCC email address (which I'm not one of them either).

It must have hung on to my work email address from some tests I used it in.  My plan is to delete and redo all five of my SBE actions and see if that stops the erroneous emails from coming to my inbox.  At first it was OK to get them because it gave me instant visibility to the fact that the flow and the SBE actions were working as designed.  But, after a while...ugh...

Rick

**Comment by BaileyAnn** [comment]
I ran across this same issue today after having to modify one of my flows. 
The workaround of removing the action and re-adding did not work for me. I also tried assigning an empty string to a variable to use in either the CC or BCC sections, but that didn't work either.


I will be trying to use the bogus email solution for now. 


Here is the error I'm getting, same as the others: 

`SendBCCthisOneEmailAddress = {!brEmailAddress} (bailey.rudd@esdec.com)
SendCCthisOneEmailAddress = {!EmptyStriing} ()
SendTOthisOneEmailAddress = {!AUemailaddresslist} (email1@ironridge.com)
subject = {!flsubject} (Design Request - Rajani Ref-7510)

Error Occurred: An Apex error occurred: SendBetterEmail.InvocableActionException: No Messages were sent. First Error: nullError Code:INVALID_EMAIL_ADDRESS - Email address is invalid:`

**Comment by jdpond** [comment]
@Rick131313 THANK YOU!  I have been trying to reproduce this without success.  Been pulling my hair out - hope it works!
